Prune transport error codes

martinthomson opened the pull request:
This prunes the transport error codes to the short list I shared on the [mailing list](https://mailarchive.ietf.org/arch/msg/quic/xJ5Rr1_N6JI1DT1S7z77G0nZG2Y).

@janaiyengar this rots quickly, I want to close on before we next chat.

Closes #96, #177, #184, #211.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -2575,7 +2576,7 @@ affected stream.
 Stream 1 is critical to the functioning of the entire connection.  If stream 1
 is closed with either a RST_STREAM or STREAM frame bearing the FIN flag, an
-endpoint MUST generate a connection error of type QUIC_CLOSED_CRITICAL_STREAM.
+endpoint MUST generate a connection error of type PROTOCOL_VIOLATION.
```
Why remove this error? It's strictly more information, and is surely useful for debugging on the side that's likely messing up.

martinthomson replied:
See comment below.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -2611,173 +2612,74 @@ CONNECTION_CLOSE or RST_STREAM frame. Error codes share a common code space.
 
-QUIC_NETWORK_IDLE_TIMEOUT (0x80000019):
-: The connection timed out due to no network activity.
+ENHANCE_YOUR_CALM (0x80000002):
```
Where's this intended to be used? It's not clear to me what exactly the endpoint wants the peer to reduce generation of -- Data? Acks? Window updates? -- and what the peer is expected to do when this error code is received. Is this a stream or connection error? It would be an odd time to say reduce generation  of something if used as a connection error... the connection is now closed. Does this error code mean that the peer SHOULD NOT reconnect?

ianswett replied:
Agreed, this does not seem helpful.

martinthomson replied:
See the analogous code in h2.  There isn't a whole lot of text in QUIC for this, but you can find the sorts of applications we use this for in h2.

ianswett replied:
If it's already in H2, should it be in the http mapping?  I can't find a good link for H2's usage, the best I found was it's informal use by twitter: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes

And why is it not error code 420?

martinthomson replied:
My understanding is that it is part of the transport in h2, not the application.  If you disagree with that assessment, let me know.

It was 420, for a while, but then we decided that a contiguous error space was more valuable.  However, it's *not* a status code.

janaiyengar replied:
If this has a specific meaning, I would want to name it as such. "ENHANCE_YOUR_CALM" is not a helpful name... but I still don't understand what it stands for. What is this used for in HTTP/2?

martinthomson replied:
Better that you search the RFC for the code and read it yourself.

janaiyengar replied:
I don't think I said anything to warrant a rude response. Perhaps you didn't mean it to be rude, but it sure sounds it.

I know of the error code from HTTP/2, and I looked it up before my first comment on this PR. RFC 7540 is not helpful in answering the questions I'm asking.

martinthomson replied:
I apologize, I just think that me trying to restate its purpose wouldn't really help because it's got a lot of context attached to it.

ianswett replied:
link?

ianswett replied:
Still confusing, particularly on a transport level.

And if we're going to add it, we're going to need to restate it's purpose, we can't just assume readers know what it means.

seanturner replied:
From RFC 7540:

    ENHANCE_YOUR_CALM (0xb):  The endpoint detected that its peer is
    exhibiting a behavior that might be generating excessive load.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -2611,173 +2612,74 @@ CONNECTION_CLOSE or RST_STREAM frame. Error codes share a common code space.
-: QUIC timed out after too many RTOs.
+: An endpoint received a frame that was badly formatted.  For instance, an empty
+  STREAM frame that omitted the FIN flag, or an ACK frame that has more
+  acknowledgment ranges than the remainder of the packet could carry.
```
These are very different errors in terms of what a peer would do with this information. I don't think collapsing these two examples is a good idea, since the peer can actually do something concrete with this information -- separating STREAM_FRAME_FORMAT_ERROR from ACK_FRAME_FORMAT_ERROR is a useful for monitoring and debugging.

ianswett replied:
+1

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -2522,13 +2520,12 @@ An endpoint MUST NOT send data on a stream at or beyond the final offset.
 
 Once a final offset for a stream is known, it cannot change.  If a RST_STREAM or
 STREAM frame causes the final offset to change for a stream, an endpoint SHOULD
```
Let's fix/discuss that separately.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -2565,6 +2562,10 @@ risks a peer missing the first such packet.  The only mechanism available to an
 
+An endpoint that receives an invalid error code in a CONNECTION_CLOSE frame MUST
+NOT signal the existence of the error to its peer.  It MAY treat the error as an
+INTERNAL_ERROR on the basis that there is some fault with the peer.
```
Yeah, this needs work.  It's trying to say that you can't respond to a CONNECTION_CLOSE with a PROTOCOL_VIOLATION.  Give me another shot at this.

janaiyengar replied:
Ah, ok, that makes sense.

martinthomson commented:
Gah, this is already rotten :(

I think that you might have misunderstood the approach that I've taken.  The point is to identify classes of errors that can have operational practices associated with them.  Finer grained errors are certainly useful during the development of a protocol, but I don't know how someone would deploy an operational practice to distinguish between a cases where it sent a badly formatted STREAM frame vs. a badly formatted ACK frame.  Those errors are going to need someone to sit down with traces and work out what is going on.

I'm not suggesting that an implementation couldn't generate different error codes for local consumption.  That's great for debugging.  But we're talking protocol signals here.  And when we use a different code point in a protocol, we have to have something in those signals that can be acted upon differently.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -675,9 +675,9 @@ increase by at least one after sending any packet.
-(connection termination is described in {{termination}}.)
+reaches 2^64 - 1, the sender MUST close the connection without sending a
+CONNECTION_CLOSE frame or any further packets; the sender MAY send a Public
+Reset packet in response to further packets that it receives.
```
Shouldn't it use the last packet number to send a connection close?

martinthomson replied:
Well, that is a possibility, but think about what it will take to reach this number.  We could put lots of effort into dealing with this case, but you have to ask yourself if that's worth it.

ianswett replied:
Ok, fair enough.  I'm happy handling that edge case in whatever way is easiest.

janaiyengar replied:
Yeah, I agree. I also realized that this max packet number is known to the receiver, and the receiver knows that the sender is near the max, so clearly it knows that the sender would close the connection.

ianswett commented:
Generating different error codes for local consumption is not a direction I'd like to go in if I can avoid it.

We have a huge error code space available, so as long as it's clear when one should use an error code, I'm not sure why we wouldn't want a relatively complete set?

janaiyengar commented:
It is very helpful operationally to have these codes available, since a server, for instance, can monitor the different error codes that are received. A spike in FRAME_ERROR for a STREAM frame is much easier to diagnose than a general FRAME_ERROR. As to why this may be useful later and not just during development -- any time you touch code, you can inadvertently introduce bugs. A server that does not want to monitor does not have to, but having the error codes makes it possible for interested parties to do so.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -2858,173 +2858,74 @@ CONNECTION_CLOSE or RST_STREAM frame. Error codes share a common code space.
 Some error codes apply only to either streams or the entire connection and have
 no defined semantics in the other context.
 
-QUIC_INTERNAL_ERROR (0x80000001):
```
you'll want this or some equivalent.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -2858,173 +2858,74 @@ CONNECTION_CLOSE or RST_STREAM frame. Error codes share a common code space.
-QUIC_INVALID_STREAM_DATA (0x8000002E):
-: STREAM frame data is malformed.
-
-QUIC_UNENCRYPTED_STREAM_DATA (0x8000003D):
```
We definitely want this.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -2858,173 +2858,74 @@ CONNECTION_CLOSE or RST_STREAM frame. Error codes share a common code space.
-QUIC_INVALID_CONNECTION_CLOSE_DATA (0x80000007):
-: CONNECTION_CLOSE frame data is malformed.
-
-QUIC_INVALID_GOAWAY_DATA (0x80000008):
```
I do want an error code per frame type.  It really helps narrow down issues.

martinthomson commented:
Discussed in Paris; we're going to add the frame-specific codes in *addition* to the generic one.

We'll remove ENHANCE_YOUR_CALM from this one.

mikkelfj commented on draft-ietf-quic-transport.md:
```diff
@@ -1213,7 +1213,7 @@ server MUST check that it would have sent a version negotiation packet if it had
 accepted the version included in the initial_version and the value differs from
 the value of negotiated_version, the server MUST terminate the connection with a
-QUIC_VERSION_NEGOTIATION_MISMATCH error.
+VERSION_NEGOTIATION_ERROR error.
```
I am having some trouble understanding how to send this error https://github.com/quicwg/base-drafts/issues/597 - but it might be resolved with updated PUBLIC_RESET to replace CONNECTION_CLOSE.

martinthomson commented:
@janaiyengar, I've rebased and updated this with what we agreed at the interim.  There is now a generic frame format error, and a set of frame-type-specific errors.  I removed ENHANCE_YOUR_CALM.

I went over the error codes that you added back in #521 and I'm going to suggest that we separate those out.  Some depend on the conclusion of #608, and I *think* that the others aren't useful to signal.  I'm sure that I'm wrong about at least one, but we should discuss them separately.

janaiyengar commented:
I'm not sure which ones from #521 are not added in here. I wouldn't remove ones that we need discussion on... If you want to discuss ones that are in #521, please add them back in and have a separate issue on removing them.
